AP_BattMonitor: Add Param for I2C_address, Move Analog & I2C Params to VarPtr Groups

hendjoshsr71 opened the pull request:
This adds a parameter for the I2C_address for smbus batteries. It also moves it and the I2C_bus into a param subgroup. The subgroup part was suggested by @WickedShell

~~I'm marking this draft to get feedback on how to do the parameter conversion to keep the old behavior (if deemed necessary)?~~

It was tested with my smartbatt PR in sitl. And included testing parameter conversion. 

New
- Adds a parameter for the I2C_address for smbus batteries
- Moves the analog specific params to a VarPtr subgroup to hide the params from other backend types
- Moves the smbus I2C params to a varptr subgroup

WickedShell commented:
Sorry I should have provided more guidance on the subtable thing. The ideal here is to actually add the subtable when the battery monitor is allocated. This allows drivers to have their own driver specific parameters. You can see an example of this in AP_RangeFinder, where the WASP driver add's it's own parameter table that only shows up if it's a WASP rangefinder selected for the type. If we do that here we wouldn't be sending SMBus related parameters to the GCS on non SMBus batteries (and similarly we could not send the analog related parameters for the SMBus ones).

hendjoshsr71 commented:
Cool! Ok I think I got it closer to what what you mentioned. 

When pulling out the analog parameters I took a guess at what it would should look like.

It is does work except for the most important part of keeping params from old to new for users! Any hints for that bit would be appreciated?

I changed the Fuel_* monitors to derive from Analog since they need some of the analog params and it seemed natural to do that?

Also I'm not sure why the smart_batt autotest doesn't want to set the params, but I can in mission planner.

hendjoshsr71 commented:
@WickedShell  I think this is ready for review now. 

The dynamic groups worked as expected and "hide" params not related.
One user note is that after the conversion a user will need to reboot in order to get their I2C battmons to work.

I have tested the parameter conversion using sitl with 9 battmons: a couple analog all with unique values, 3 different smbus monitors, and fuel flow monitors. 

I edited the autotest as it would not work without additional resets now that the smbus params are dynamically allocated.

The build size increase seems worth the cost for custom smbus addresses and cleaning up params for backends 200B for main vehicle types. However, Matek405 increases in size mainly because NeoDesign backends aren't behind the `#if HAL_BATTMON_SMBUS_ENABLE`

hendjoshsr71 commented:
rebased

tridge commented on libraries/AP_BattMonitor/AP_BattMonitor.cpp:
```diff
@@ -215,7 +310,48 @@ AP_BattMonitor::init()
+void AP_BattMonitor::convert_dynamic_param_groups(uint8_t instance)
+{
+    AP_Param::ConversionInfo info;
+    if (!AP_Param::find_top_level_key_by_pointer(this, info.old_key)) {
```
start with looking up the BATTn_TYPE param, and offset from that would be safer

hendjoshsr71 replied:
Done. I did try to find the index using pointers instead , but no appropriate functions exists to do that I could find. I did try modifying the existing find_key_by_group to do that but the structure of the param table eludes me to generalize the pointer math.

tridge reviewed (changes requested):
it would be good to valgrind this

hendjoshsr71 commented:
Addressed comments about indexing from `BattN_Monitor`
Tested through each vehicle type with Valgrind enabled, using unique values for each BattMon type being changed and parameters being moved. (Did find something with Valgrind on Sub, fix in this PR https://github.com/ArduPilot/ardupilot/pull/18094)

Confirmed param conversion worked for all monitors below.
After the parameter conversion occurred, checked values could be changed and that those values were kept on a reboot
- 3 Analog 
- 1 Rotoye, 1 Maxell, 1 SMBus_Generic
- 1 fuel flow
- 1 fuel flow PWM
- 1 CAN (no change happening on this one)

hendjoshsr71 commented:
add to I2C_ADDR and I2C_BUS `@RebootRequired: True`

hendjoshsr71 commented:
To do: figure out the flash cost of this PR and where it comes from for F405 boards.
